core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -414,6 +414,16 @@ public PlanWithProperties visitTopN(TopNNode node, PreferredProperties preferred
                     break;
                 case PARTIAL:
                     child = planChild(node, PreferredProperties.any());
+                    // If source is pre-sorted, partial topN can be replaced with partial Limit N
+                    List<LocalProperty<Symbol>> desiredProperties = node.getOrderingScheme().getOrderBy().stream()
+                            .map(symbol -> new SortingProperty<>(symbol, node.getOrderingScheme().getOrdering(symbol)))
+                            .collect(toImmutableList());
+                    if (LocalProperties.match(child.getProperties().getLocalProperties(), desiredProperties).stream()
+                            .noneMatch(Optional::isPresent)) {
+                        return withDerivedProperties(
+                                new LimitNode(node.getId(), child.getNode(), node.getCount(), true),
+                                child.getProperties());
```

Comment:
I am not sure i can reason about this (@martint or @dain would do better), but i noticed the same condition being used here, in this class for visitSort 

https://github.com/trinodb/trino/blob/1b7787d4f91d5dc1e935e68507432922a4cbf210/core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java#L429-L442

thoughts

- in sort case, we do this only on `isSingleNode` condition
- we must verify that exposing bucket sort by property from Hive to optimizer does not eliminate Sort -- that wouldn't be correct
- the LimitNode inserted here can be separated from TS and the data can come mixed, eg merged from multiple splits
  - that's my hypothesis, i may be wrong
  - i think current plan is considered as not parallelized, and it's parallelized in `AddLocalExchanges`, so it may seem i am overly concerned ...
  - ...  but then declaring Hive table as 'sorted' is not correct -- the TS data comes from multiple files, multiple stripes, in no particular order, and so it's not actually sorted

Reply:
> we must verify that exposing bucket sort by property from Hive to optimizer does not eliminate Sort -- that wouldn't be correct

It would be correct if sort is running on single node, then local stream would become global stream (as comment there says).

> ... but then declaring Hive table as 'sorted' is not correct -- the TS data comes from multiple files, multiple stripes, in no particular order, and so it's not actually sorted

Inserted limit is partial as `TopNNode` was.

> the LimitNode inserted here can be separated from TS and the data can come mixed, eg merged from multiple splits 

Limit can be pushed down, but I don't think sorting property would be broken then. It would be incorrect (and pre-existing bug) to pushdown limit when sorting property is relevant. CC @martint

Reply:
> ... but then declaring Hive table as 'sorted' is not correct -- the TS data comes from multiple files, multiple stripes, in no particular order, and so it's not actually sorted
> 
> Inserted limit is partial as TopNNode was (per sorted stream)

I'm not sure how it works but sure we cannot replace partial topN with partial limit on unordered data.

Reply:
> I'm not sure how it works but sure we cannot replace partial topN with partial limit on unordered data.

Yes. The code here ensures that data is ordered within stream.